extension/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+
+        let mut duration_to_start = (earliest_time - self.current_window_min) as f64
+            / USECS_PER_SEC as f64;
+        let duration_to_end = (self.current_window_max - latest_time)  as f64
+            / USECS_PER_SEC as f64;
+
+        let sampled_interval = (latest_time - earliest_time) as f64
+            / USECS_PER_SEC as f64;
+        let avg_duration_between_samples = sampled_interval as f64
+            / (self.window.len()-1) as f64;
+
+        if self.is_counter && result_val > 0.0 && earliest_val >= 0.0 {
+            // Counters cannot be negative. If we have any slope at
+            // all (i.e. result_val went up), we can extrapolate
+            // the zero point of the counter. If the duration to the
+            // zero point is shorter than the durationToStart, we
+            // take the zero point as the start of the series,
+            // thereby avoiding extrapolation to negative counter
+            // values.
+            let duration_to_zero = sampled_interval * (earliest_val / result_val);
+            if duration_to_zero < duration_to_start {
+                duration_to_start = duration_to_zero
+            }
+        }
+
+        // If the first/last samples are close to the boundaries of the range,
+        // extrapolate the result. This is as we expect that another sample
+        // will exist given the spacing between samples we've seen thus far,
+        // with an allowance for noise.
+
+        let extrapolation_threshold = avg_duration_between_samples * 1.1;
+        let mut extrapolate_to_interval = sampled_interval;
+
+        if duration_to_start < extrapolation_threshold {
+            extrapolate_to_interval += duration_to_start
+        } else {
+            extrapolate_to_interval += avg_duration_between_samples / 2.0
+        }
+
+        if duration_to_end < extrapolation_threshold {
+            extrapolate_to_interval += duration_to_end
+        } else {
+            extrapolate_to_interval += avg_duration_between_samples / 2.0
+        }
+
+        result_val = result_val * (extrapolate_to_interval / sampled_interval);
+
+        if self.is_rate {
+            //TODO is this right?
+            result_val = result_val
+                / ((self.current_window_max - self.current_window_min) /USECS_PER_SEC) as f64
+        }
+
+        self.deltas.push(result_val.to_datum());
+        self.nulls.push(false);
+    }
+
+    pub fn to_pg_array(&mut self) -> *mut ArrayType {
+        while self.current_window_max <= self.greatest_time {
+            self.flush_current_window();
+        }
+        unsafe {
+            construct_md_array(
+                self.deltas.as_mut_ptr(),
+                self.nulls.as_mut_ptr(),
+                1,
+                &mut (self.nulls.len() as _),
+                &mut 1,
+                FLOAT8OID,
+                size_of::<f64>() as _,
+                FLOAT8PASSBYVAL != 0,
+                'd' as u8 as _,
```

Comment:
what does this `as _` do

Reply:
converts it to whatever type is expected in that context, in this case the first one casts from `usize` to postgres's `Size` and the second one casts from `u8` to `c_char`